src/core/thread/discover_scanner.cpp

```diff
@@ -225,6 +231,12 @@ void DiscoverScanner::HandleDiscoveryRequestFrameTxDone(Message &aMessage)
 
 void DiscoverScanner::HandleDiscoverComplete(void)
 {
+    // Restore Data Polling or CSL for rx-off-when-idle device.
+    if (!Get<Mle>().IsDisabled() && !Get<Mle>().IsRxOnWhenIdle())
+    {
+        Get<MeshForwarder>().SetRxOnWhenIdle(false);
+    }
```

Comment:
@LuDuda one situation that we may want to consider:
- What happens if device supports and does CSL (rx) and CSL window happens in middle of scan.
- Here we change `RxOnWhenIdle` on `MeshForwarder` and `Mac` but keep it the same on `Mle`.
- The `Mac::IsCslEnabled()` checks with `Mle::IsRxOnWhenIdle()` so should continue to think CSL Is enabled.
- I expect with current code, we stay in rx mode on the scan channel in this case (even though `SubMac` still thinks CSL is enabled) - though I am not 100% sure.

Main question is what is the better/desired behavior? 
1. Should we ignore CSL window and stay in scan channel?
2. Should we ignore scan and do CSL sample (on CSL channel)?

I slightly prefer 1 (seems ok to miss CSL window during scan and I feel would be simpler to implement as well).

Since this is kind of tricky, may be good to cover this with a test-case (to make sure it all works fine)?

Reply:
I think option 1 has indeed more sense.

I haven't seen any conflicts between CSL and scanning in the simulation platform as [here](https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L1167) we skip CSL sampling if we are in TX or RX states, but I do [see](https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L1162https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L1162) that for some devices it won't work exactly like this. 

Do you have some recommendations on how to tackle this? I tested that adding the following check [here](https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L1162) `(mState != kStateDisabled) && (mState != kStateReceive))` seems to work fine on Nordic devices i.e. scanning has higher priority than CSL sampling. Another potential option would be to pass the `mRxOnWhenIdle` variable down to SubMac and exit [here](https://github.com/openthread/openthread/blob/main/src/core/mac/sub_mac.cpp#L1161) if the variable is set.

Reply:
Thanks @LuDuda.

I think we have two cases depending on `RadioSupportsReceiveTiming()` value.
- If it does not support, I agree with you that things should work
- If it does, then where we call `ReceiveAt()` which can make things tricky.

Some ideas:

A) We can change `Mac::IsCslEnabled()` as `return !Get<Mle::Mle>().IsRxOnWhenIdle() && !mRxOnWhenIdle && IsCslCapable();`
   - Basically stop CSL during scan
   - This feels safe (in terms of behavior)
   - The major downside is after scan is done we would start a "Child Update Request" exchange from `Mac::UpdatseCsl()` (which we wanted to avoid, otherwise we could change mode during scan).
  
B) Let `ReceiveAt()` happen
  - Which I think means we prefer CSL over scan rx in this case
  - The downside is I think `otPlatRadioReceiveAt()` after the duration is over, will put radio to sleep (we want to go back to rx on current scan channel)

C) Pass `mRxOnWhenIdle` down
- The downside is how to handle RCP case (may need new spinel prop and deal with compatibility issue with host/RCP)

D) Add the check `(mState != kStateReceive)` as you suggested in `HandleCslTimer` before calling `ReceiveAt()`
- This seems reasonable to me
- The only downside I can think of for this is that data poll rx window may be preferred over CSL rx (if we are in rx after data poll to rx from parent, we skip calling `ReceiveAt()`). I think this may be even better behavior?

I would say D seems to be the best option.

I will think more on this to see if can do something else.

Reply:
I've rebased and Implemented option D.

@edmont could you please check this [line](https://github.com/openthread/openthread/pull/8832/files#diff-87a23b12bc8de7d17487413e94327bc95169ee65a44c3a18fb0905c87e708657R1164) before we potentially merge this change.

Reply:
Thanks.  BTW, I could not really find any better options than `D`.